### PR TITLE
NEW SingleRecordAdmin class for editing one record at a time

### PR DESCRIPTION
**Pull request by GuySartorelli**
Note that there may be some methods still on `LeftAndMain` that will be migrated into this class, such as `save()` which isn't used by `CMSMain` or `ModelAdmin` - but moving those will be handled by https://github.com/silverstripe/silverstripe-admin/issues/1763 once the bulk of the other refactoring has been completed. This is to reduce the chance of regressions in areas that still need to be refactored.

## TODO

1. CMS 5 PR to deprecate `SiteConfigLeftAndMain::save_siteconfig()` and `LeftAndMain.tree_config`

## Issue
- https://github.com/silverstripe/silverstripe-admin/issues/1764

### CODE REVIEW

File: code/CMSProfileController.php [hunk 1]
```diff
@@ -2,113 +2,75 @@
...
+class CMSProfileController extends SingleRecordAdmin
 {
     private static $url_segment = 'myprofile';
 
     private static $menu_title = 'My Profile';
 
     private static $required_permission_codes = 'CMS_ACCESS';
 
-    private static $tree_class = Member::class;
+    private static $model_class = Member::class;
```
Comment (ChloeSartorelli) [inline code/CMSProfileController.php:19]: See change in LeftAndMain

File: code/LeftAndMain.php [hunk 2]
```diff
@@ -99,7 +99,7 @@ class LeftAndMain extends AdminController implements PermissionProvider
      * @config
      * @var string
      */
-    private static $tree_class = null;
+    private static $model_class = null;
```
Comment (ChloeSartorelli) [inline code/LeftAndMain.php:103]: Most subclasses of `LeftAndMain` which use this config are _not_ trees - and even with `CMSMain` which is the only tree structure, this is getting refactored to be more generic anyway.

Makes sense to update this to be named something that _all_ the classes using it can respect.

File: code/LeftAndMain.php [hunk 3]
```diff
@@ -1223,24 +1223,7 @@ public function getEditForm($id = null, $fields = null)
...
-        } elseif ($record->hasMethod('getCMSValidator')) {
-            // BC support for framework < v4.7
-            $validator = $record->getCMSValidator();
-
-            // The clientside (mainly LeftAndMain*.js) rely on ajax responses
-            // which can be evaluated as javascript, hence we need
-            // to override any global changes to the validation handler.
-            if ($validator) {
-                $form->setValidator($validator);
-            }
```
Comment (ChloeSartorelli) [inline code/LeftAndMain.php:1238]: All of this boils down to `$form->setValidator($record->getCMSCompositeValidator())`, because all `DataObject` classes have that method, and `$record` will always be a `DataObject`.

File: code/LeftAndMain.php [hunk 4]
```diff
@@ -1223,24 +1223,7 @@ public function getEditForm($id = null, $fields = null)
...
-        } else {
-            $form->unsetValidator();
-        }
```
Comment (ChloeSartorelli) [inline code/LeftAndMain.php:1241]: That method just sets it null anyway so even if somehow `getCMSCompositeValidator()` returned null (which it can't) the result would be the same.

File: code/CMSProfileController.php
Comment (ChloeSartorelli) [inline code/CMSProfileController.php:1]: Anything removed from this class was either not necessary (likely related to legacy CMS 3 or older code that no longer exists) or is already being done in one of the superclasses.

File: code/CMSProfileController.php [hunk 5]
```diff
@@ -2,113 +2,75 @@
...
-        }
+    private static bool $allow_new_record = false;
 
-        if ($form->Fields()->hasTabSet()) {
-            $form->Fields()->findOrMakeTab('Root')->setTemplate('SilverStripe\\Forms\\CMSTabSet');
+    protected function init()
+    {
+        parent::init();
+        if (!$this->getResponse()->isRedirect()) {
+            $this->setCurrentPageID(Security::getCurrentUser()->ID);
```
Comment (ChloeSartorelli) [inline code/CMSProfileController.php:31]: Moving this into `init` instead of `getEditForm` so it takes affect for _all_ actions on this controller (including any that might be added through extensions).

Skipping for redirects because at that point we're not doing anything in this controller anyway.

File: code/SecurityAdmin.php [hunk 6]
```diff
@@ -58,7 +58,7 @@ class SecurityAdmin extends ModelAdmin implements PermissionProvider
 
     private static $menu_priority = 0;
 
-    private static $tree_class = Group::class;
+    private static $model_class = Group::class;
```
Comment (ChloeSartorelli) [inline code/SecurityAdmin.php:61]: I strongly suspect this config doesn't actually do anything for this class, but checking that is out of scope for this PR. I'm just updating it to the config that `LeftAndMain` expects.

File: code/SingleRecordAdmin.php [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+ */
+abstract class SingleRecordAdmin extends LeftAndMain
+{
+    /**
+     * Determines if there should be a single record in the database that this admin edits.
+     *
+     * If this is not true, you need to provide a mechanism to tell the form which record it should be editing.
+     * This could be an action (e.g. edit/$ID), or could be based on something about the current member, etc.
+     */
+    private static bool $only_one_record = true;
```
Comment (ChloeSartorelli) [inline code/SingleRecordAdmin.php]: This is for `SiteConfig` style functionality - but since `CMSProfileController` is explicitly for the _current_ member and not _the one member in the database_, it needs to be configurable.

File: code/SingleRecordAdmin.php [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+    /**
+     * If no record exists, allow a new one to be created
+     */
+    private static bool $allow_new_record = true;
```
Comment (ChloeSartorelli) [inline code/SingleRecordAdmin.php:24]: For almost all controllers, we'll want to be able to create a record if there isn't one. `CMSProfileController` is the exception. Even with `SiteConfig`, if the record in the DB gets somehow deleted, you'll want to be able to create a new one.

File: code/LeftAndMain.php (hunk 2 above)
Comment (emteknetnz) [reply code/LeftAndMain.php:102]: Is `$model_class` always a `DataObject`? If so this should probably be called `$data_class` instead which as far as I can tell the most common way refer to the "class of the DataObject" i.e. $dataClass.  $model_class seems like it could be any model, including something that's not a DataObject

Comment (emteknetnz) [reply code/LeftAndMain.php:103]: `$model_class` is always a DataObject correct? If so this should probably be called `$data_class` instead which as far as I can tell the most common way refer to the "class of the DataObject" i.e. `$dataClass`. `$model_class` seems like it could be any model, including something that's not a `DataObject`

File: code/SingleRecordAdmin.php [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+ */
+abstract class SingleRecordAdmin extends LeftAndMain
+{
+    /**
+     * Determines if there should be a single record in the database that this admin edits.
+     *
+     * If this is not true, you need to provide a mechanism to tell the form which record it should be editing.
+     * This could be an action (e.g. edit/$ID), or could be based on something about the current member, etc.
+     */
+    private static bool $only_one_record = true;
```
Comment (emteknetnz) [inline code/SingleRecordAdmin.php]:
```suggestion
    private static bool $single_record = true;
```
`$only_one_record` sounds weird to me, almost implying that it's a bad thing :-)

File: code/SingleRecordAdmin.php (hunk 8 above)
Comment (emteknetnz) [reply code/SingleRecordAdmin.php:24]: Seems like these two new configurable properties shouldn't be configurable? I can't think of a scenario where they would be better of simply being protected non-static properties instead? There's a miniscule performance benefit, though more importantly it prevents people from doing things they shouldn't via config, like changing the the config values of SecurityAdmin

File: code/SingleRecordAdmin.php [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+
+/**
+ * The base class for admin sections that exist for editing a single record.
+ */
+abstract class SingleRecordAdmin extends LeftAndMain
+{
+    /**
+     * Determines if there should be a single record in the database that this admin edits.
+     *
+     * If this is not true, you need to provide a mechanism to tell the form which record it should be editing.
```
Comment (emteknetnz) [inline code/SingleRecordAdmin.php]:
```suggestion
     * If false, you need to provide a mechanism to tell the form, usually via getEditForm(), which record it should be editing.
```

File: code/CMSProfileController.php [hunk 11]
```diff
@@ -2,113 +2,75 @@
...
         }
-        $origLocale = $member->Locale;
-
-        if (!$member->canEdit()) {
-            $form->sessionMessage(_t(__CLASS__.'.CANTEDIT', 'You don\'t have permission to do that'), 'bad');
-            return $this->redirectBack();
-        }
-
+        // Get the current member locale so we can see if it changes
+        $member = Member::get()->byID($data['ID']);
```
Comment (emteknetnz) [inline code/CMSProfileController.php]:
```suggestion
        $member = Member::get()->byID($id);
```

File: code/CMSProfileController.php [hunk 12]
```diff
@@ -2,113 +2,75 @@
...
     }
 
     public function save(array $data, Form $form): HTTPResponse
     {
-        $member = Member::get()->byID($data['ID']);
-        if (!$member) {
-            $this->httpError(404);
+        // Make sure the ID is a positive number
+        $id = $data['ID'] ?? null;
+        if (!is_numeric($id) || $id < 1) {
```
Comment (emteknetnz) [inline code/CMSProfileController.php]: Should validate that it's an int (or string int), is_numeric() will allow a float

File: code/SingleRecordAdmin.php [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+
+    public function getRecord($id): ?DataObject
+    {
+        if (!$id) {
+            return $this->getSingleRecord();
+        }
+        return parent::getRecord($id);
+    }
+
+    protected function getSingleRecord(): ?DataObject
```
Comment (emteknetnz) [inline code/SingleRecordAdmin.php:44]:
```suggestion
    private function getSingleRecord(): ?DataObject
```
If we're providing `only_one_record` and `allow_new_record` and implementing logic to do stuff based on how they're set, why would anyone need to override this in a subclass? Things should be private unless there's an immediate need to make it protected.

File: code/SingleRecordAdmin.php [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+        }
+        return $form;
+    }
+
+    public function getRecord($id): ?DataObject
+    {
+        if (!$id) {
+            return $this->getSingleRecord();
+        }
+        return parent::getRecord($id);
```
Comment (emteknetnz) [inline code/SingleRecordAdmin.php:41]: What's the scenario where $id is truthy?  Seems like we should be throwing an exception here instead?

**Review by emteknetnz — CHANGES_REQUESTED** [review]

File: code/LeftAndMain.php (hunk 2 above)
Comment (ChloeSartorelli) [reply code/LeftAndMain.php:103]: I chose this name to complement `getModelClass()` which currently only exists in `ModelAdmin` but will become more ubiquitous in other cards in the epic. 

In cmsmain and likely others the `getModelClass()` method will return the value held in this config.

If you're okay with `getModelClass()` returning the value of a `data_class` config property I can make the requested change - though that would be a little unintuitive.

File: code/SingleRecordAdmin.php (hunk 8 above)
Comment (ChloeSartorelli) [reply code/SingleRecordAdmin.php:24]: The pattern in these controllers is to use configuration for this sort of thing, so I was just following that pattern.
If you feel strongly about it, I can change them to protected member properties. 

`SecurityAdmin` isn't relevant, I assume you mean the profile admin. People can do what they like with it in their own projects.

File: code/SingleRecordAdmin.php (hunk 9 above)
Comment (ChloeSartorelli) [reply code/SingleRecordAdmin.php]: Feel free to suggest an alternative name.

File: code/SingleRecordAdmin.php (hunk 10 above)
Comment (ChloeSartorelli) [reply code/SingleRecordAdmin.php]: Will make the change, but won't add "usually via getEditForm()" because the only scenario we've actually implemented doesn't do it that way.

**Comment by GuySartorelli** [comment]
I've responded to the bits that are relevant for the CMS 5 PRs. I'll make the requested changes that are agreed upon after the CMS 5 PRs have been merged and I can rebase on top of them.

File: code/LeftAndMain.php (hunk 2 above)
Comment (emteknetnz) [reply code/LeftAndMain.php:103]: Yeah, I guess ModelAdmin does use $modelClass everywhere right now, so I guess if that's the local standard we should be using that

File: code/SingleRecordAdmin.php (hunk 9 above)
Comment (emteknetnz) [reply code/SingleRecordAdmin.php]: $single_record

Comment (ChloeSartorelli) [reply code/SingleRecordAdmin.php]: How about `only_single_record`? Or `restrict_to_single_record`? I want the name to be fairly clear that it's saying "there should only ever be one record of this type" when set to true.

Or maybe `dynamic_current_record`? As in "this will dynamically fetch the current record without you telling it what the ID is"?

File: code/SingleRecordAdmin.php (hunk 8 above)
Comment (emteknetnz) [reply code/SingleRecordAdmin.php:24]: Yup, I meant CMSProfileController

You're probably right it is something of a pattern, which it probably shouldn't be though we should respect what's already there.

File: code/SingleRecordAdmin.php (hunk 9 above)
Comment (emteknetnz) [reply code/SingleRecordAdmin.php]: $restrict_to_single_record sounds good

Comment (ChloeSartorelli) [reply code/SingleRecordAdmin.php]: Done

File: code/SingleRecordAdmin.php (hunk 10 above)
Comment (ChloeSartorelli) [reply code/SingleRecordAdmin.php]: Done

File: code/CMSProfileController.php (hunk 11 above)
Comment (ChloeSartorelli) [reply code/CMSProfileController.php]: Done

File: code/CMSProfileController.php (hunk 12 above)
Comment (ChloeSartorelli) [reply code/CMSProfileController.php]: Done

File: code/SingleRecordAdmin.php (hunk 14 above)
Comment (ChloeSartorelli) [reply code/SingleRecordAdmin.php:41]: `$id` is truthy if an ID is passed in. So we either want the "single record" based on the configuration (where a `null` or `0` is passed in), or we want the specified record if a specific ID is passed in.

File: code/SingleRecordAdmin.php (hunk 13 above)
Comment (ChloeSartorelli) [reply code/SingleRecordAdmin.php:44]: Please see the docs where overriding this method is explicitly mentioned as a way to provide more complex behaviour for getting this record.

File: code/SingleRecordAdmin.php (hunk 14 above)
Comment (emteknetnz) [reply code/SingleRecordAdmin.php:41]: Why would be specify an $id though in a SingleRecordAdmin? Isn't the whole point to only have a single record?

**Review by emteknetnz — APPROVED** [review]

**Review by emteknetnz — APPROVED** [review]
